website/spam_admin/utils.py

```diff
@@ -0,0 +1,195 @@
+        'author': {
+            'email': project.creator.emails,
+            'name': project.creator.fullname,
+        },
+        'pid': project._id,
+        'components': [serialize_project(component) for component in project.nodes]
+    }
+
+def _format_spam_node_data(node):
+    from website.addons.wiki.model import NodeWikiPage
+    from website.views import serialize_log
+
+    logs = []
+    for log in reversed(node.logs):
+        if log:
+            logs.append(serialize_log(log))
+
+    content = {
+        'wikis': [wiki.content for wiki in NodeWikiPage.find(Q('node', 'eq', node))],
+        'logs': logs,
+        'tags': [tag._id for tag in node.tags],
+        'components': str([_format_spam_node_data(component) for component in node.nodes])
+    }
+    data = {
+        'message': content,
+        'project_title': node.title,
+        'category': node.category_display,
+        'description': node.description or '',
+        'url': node.url,
+        'absolute_url': node.absolute_url,
+        'date_created': iso8601format(node.date_created),
+        'date_modified': iso8601format(node.logs[-1].date) if node.logs else '',
+        'date': iso8601format(node.logs[-1].date) if node.logs else '',
+        'tags': [tag._id for tag in node.tags],
+        'is_registration': node.is_registration,
+        'registered_from_url': node.registered_from.url if node.is_registration else '',
+        'registered_date': iso8601format(node.registered_date) if node.is_registration else '',
+        'registration_count': len(node.node__registrations),
+        'is_fork': node.is_fork,
+        'forked_from_id': node.forked_from._primary_key if node.is_fork else '',
+        'forked_from_display_absolute_url': node.forked_from.display_absolute_url if node.is_fork else '',
+        'forked_date': iso8601format(node.forked_date) if node.is_fork else '',
+        'fork_count': len(node.node__forked.find(Q('is_deleted', 'eq', False))),
+        'templated_count': len(node.templated_list),
+        'watched_count': len(node.watchconfig__watched),
+        'private_links': [x.to_json() for x in node.private_links_active],
+        'points': len(node.get_points(deleted=False, folders=False)),
+        'comment_level': node.comment_level,
+        'has_comments': bool(getattr(node, 'commented', [])),
+        'has_children': bool(getattr(node, 'commented', False)),
+        'author': node.creator.fullname,
+        'email': node.creator.emails,
+    }
+    return data
+
+def _project_is_spam(node):
+    if not settings.SPAM_ASSASSIN:
+            return False
+    try:
+        data = _format_spam_node_data(node)
+        resp = requests.post(SPAM_ASSASSIN_URL, data=json.dumps(data))
+
+        if json.loads(resp.text).get('decision', '') == "SPAM":
+            return True
+        return False
+    except:
+        return False
+
+def train_spam_project(project, is_spam):
+    if not settings.SPAM_ASSASSIN:
+            return False
+    try:
```

Comment:
We do not want the OSF to be dependent on Spam Assassin whatsoever. Spam Assassin is only an accessory. Thus, this blanket try-catch statement serves to catch any and all errors that may occur due to calling the separate Spam Assassin service. The idea here is to have this entire method fail silently if anything goes wrong.